pkg/git/libgit2/checkout.go

```diff
@@ -69,96 +69,151 @@ type CheckoutBranch struct {
+			return nil, fmt.Errorf("can't use managed transport without a valid transport auth id.")
+		}
+		managed.AddTransportOptions(opts.TransportOptionsURL, managed.TransportOptions{
+			TargetURL:    url,
+			AuthOpts:     opts,
+			ProxyOptions: &git2go.ProxyOptions{Type: git2go.ProxyTypeAuto},
+		})
+		url = opts.TransportOptionsURL
+		remoteCallBacks := managed.RemoteCallbacks()
+		defer managed.RemoveTransportOptions(opts.TransportOptionsURL)
+
+		repo, remote, err := initializeRepoWithRemote(ctx, path, url, opts)
+		if err != nil {
+			return nil, err
+		}
+		// Open remote connection.
+		err = remote.ConnectFetch(&remoteCallBacks, nil, nil)
 		if err != nil {
-			return nil, fmt.Errorf("unable to remote ls for '%s': %w", managed.EffectiveURL(url), gitutil.LibGit2Error(err))
+			remote.Free()
+			repo.Free()
+			return nil, fmt.Errorf("unable to fetch-connect to remote '%s': %w", managed.EffectiveURL(url), gitutil.LibGit2Error(err))
 		}
-		if len(heads) > 0 {
-			hash := heads[0].Id.String()
-			currentRevision := fmt.Sprintf("%s/%s", c.Branch, hash)
-			if currentRevision == c.LastRevision {
-				// Construct a partial commit with the existing information.
-				c := &git.Commit{
-					Hash:      git.Hash(hash),
-					Reference: "refs/heads/" + c.Branch,
+		defer func() {
+			remote.Disconnect()
+			remote.Free()
+			repo.Free()
+		}()
+
+		// When the last observed revision is set, check whether it is still the
+		// same at the remote branch. If so, short-circuit the clone operation here.
+		if c.LastRevision != "" {
+			heads, err := remote.Ls(c.Branch)
+			if err != nil {
+				return nil, fmt.Errorf("unable to remote ls for '%s': %w", managed.EffectiveURL(url), gitutil.LibGit2Error(err))
+			}
+			if len(heads) > 0 {
+				hash := heads[0].Id.String()
+				currentRevision := fmt.Sprintf("%s/%s", c.Branch, hash)
+				if currentRevision == c.LastRevision {
+					// Construct a partial commit with the existing information.
+					c := &git.Commit{
+						Hash:      git.Hash(hash),
+						Reference: "refs/heads/" + c.Branch,
+					}
+					return c, nil
 				}
-				return c, nil
 			}
 		}
-	}
 
-	// Limit the fetch operation to the specific branch, to decrease network usage.
-	err = remote.Fetch([]string{c.Branch},
-		&git2go.FetchOptions{
-			DownloadTags:    git2go.DownloadTagsNone,
-			RemoteCallbacks: remoteCallBacks,
-			ProxyOptions:    git2go.ProxyOptions{Type: git2go.ProxyTypeAuto},
-		},
-		"")
-	if err != nil {
-		return nil, fmt.Errorf("unable to fetch remote '%s': %w",
-			managed.EffectiveURL(url), gitutil.LibGit2Error(err))
-	}
+		// Limit the fetch operation to the specific branch, to decrease network usage.
+		err = remote.Fetch([]string{c.Branch},
+			&git2go.FetchOptions{
+				DownloadTags:    git2go.DownloadTagsNone,
+				RemoteCallbacks: remoteCallBacks,
+			},
+			"")
+		if err != nil {
+			return nil, fmt.Errorf("unable to fetch remote '%s': %w",
+				managed.EffectiveURL(url), gitutil.LibGit2Error(err))
```

Comment:
Since the passed in URL is now the original URL, we don't need to use `EffectiveURL()` here. If the `url` variable is not overwritten above, we still have the original URL.

Reply:
The passed in URL is original but it's always overwritten here: https://github.com/fluxcd/source-controller/blob/main/pkg/git/libgit2/checkout.go#L96